src/Microsoft.AspNetCore.Routing/Matching/ILEmitTrieFactory.cs

```diff
@@ -0,0 +1,588 @@
+            // of the text.
+            var groups = entries.GroupBy(e => GetUInt64Key(e.text, index));
+            foreach (var group in groups)
+            {
+                // if (uint64Value == 0x.....) { ... }
+                var next = il.DefineLabel();
+                il.Emit(OpCodes.Ldloc, locals.UInt64Value);
+                il.Emit(OpCodes.Ldc_I8, unchecked((long)group.Key));
+                il.Emit(OpCodes.Bne_Un, next);
+
+                // Process the group
+                EmitTable(il, group.ToArray(), index + 4, length, locals, labels, methods);
+                il.MarkLabel(next);
+            }
+
+            // goto: defaultDestination
+            il.Emit(OpCodes.Br, labels.ReturnDefault);
+        }
+
+        private static void EmitSingleCharacterTable(
+            ILGenerator il,
+            (string text, int destination)[] entries,
+            int index,
+            int length,
+            Locals locals,
+            Labels labels,
+            Methods methods)
+        {
+            // See the vectorized code path for a much more thorough explanation.
+
+            // IMPORTANT
+            //
+            // If you are modifying this code, be aware that the easiest way to make a mistake is by
+            // getting the set of casts wrong doing something like:
+            //
+            // il.Emit(OpCodes.Ldc_I4, ~0x007F);
+            //
+            // The IL Emit apis don't have overloads that accept ulong or ushort, and will resolve
+            // an overload that does an undesirable conversion (for instance convering ulong to float).
+            //
+            // IMPORTANT
+
+            // Unsafe.ReadUnaligned<ushort>(ref p)
+            il.Emit(OpCodes.Ldloc, locals.P);
+            il.Emit(OpCodes.Call, methods.ReadUnalignedUInt16);
+
+            // uint16Value = ...
+            il.Emit(OpCodes.Stloc, locals.UInt16Value);
+
+            // Unsafe.Add(ref p, 2)
+            il.Emit(OpCodes.Ldloc, locals.P);
+            il.Emit(OpCodes.Ldc_I4, 2); // 2 bytes were read
+            il.Emit(OpCodes.Call, methods.Add);
+
+            // p = ref ...
+            il.Emit(OpCodes.Stloc, locals.P);
+
+            // if ((uInt16Value & ~0x007FUL) == 0)
+            // {
+            //     goto: NotAscii;
+            // }
+            il.Emit(OpCodes.Ldloc, locals.UInt16Value);
+            il.Emit(OpCodes.Ldc_I4, unchecked((int)((uint)~0x007F)));
+            il.Emit(OpCodes.And);
+            il.Emit(OpCodes.Brtrue, labels.ReturnNotAscii);
+
+            // Since we're handling a single character at a time, it's easier to just
+            // generate an 'if' with two comparisons instead of doing complicated conversion
+            // logic.
+
+            // Now we generate an 'if' ladder with an entry for each of the unique
+            // characters in the group. 
+            var groups = entries.GroupBy(e => GetUInt16Key(e.text, index));
```

Comment:
Why group by ushort instead of char? Add a comment explaining why?

Reply:
I just did it for consistency. It's easier for me to think about this uint16/uint64 vs char/ulong